### PR TITLE
doc: add v0.5.x -> v0.6.0 update steps

### PR DESCRIPTION
**Pull request by fanminshi**
[skip ci]

### CODE REVIEW

**Comment by fanminshi** [comment]
added rolling upgrade steps

File: doc/user/upgrade/upgrade_guide.md [hunk 1]
```diff
@@ -22,6 +22,80 @@ $ kubectl edit deployment/etcd-operator
...
+## v0.5.x -> v0.6.0
+### Breaking Change: 
+The v0.6.0 release deprecates the support for [operator level s3 backup](https://github.com/coreos/etcd-operator/blob/v0.5.1/doc/user/backup_config.md#operator-level-configuration).
+
+### Steps to upgrade:  
+If you are using operator level s3 backup in v0.5.x, you must migrate to use [cluster level backup](../backup_config.md#s3-on-aws) in v0.6.0.
+
+Assuming your cluster uses [operator level s3 backup](https://github.com/coreos/etcd-operator/blob/v0.5.1/doc/user/backup_config.md#operator-level-configuration).
+
+- Create a backup of you current data using [backup config guide](../backup_config.md).
```
Comment (hasbro17) [inline doc/user/upgrade/upgrade_guide.md]: ... using the [backup service guide](https://github.com/coreos/etcd-operator/blob/master/doc/user/backup_service.md#http-api-v1)

File: doc/user/upgrade/upgrade_guide.md [hunk 2]
```diff
@@ -22,6 +22,80 @@ $ kubectl edit deployment/etcd-operator
...
+  ```sh
+  $ kubectl -n <namespace-name> delete configmap aws
+  $ kubectl -n <namespace-name> delete secret aws
+  ```
+
+- Create AWS secret for [cluster level backup](../backup_config.md#s3-on-aws):
+
+  `$ kubectl -n <namespace-name> create secret generic aws --from-file=$AWS_DIR/credentials --from-file=$AWS_DIR/config`
+
+- Update Cluster spec to enable [cluster level backup](../backup_config.md#s3-on-aws):
```
Comment (hasbro17) [inline doc/user/upgrade/upgrade_guide.md]: Perhaps reword a bit?

Change the `backup` field in your existing Cluster spec to enable . . .

File: doc/user/upgrade/upgrade_guide.md [hunk 3]
```diff
@@ -22,6 +22,80 @@ $ kubectl edit deployment/etcd-operator
...
+
```
Comment (hasbro17) [inline doc/user/upgrade/upgrade_guide.md]: From

File: doc/user/upgrade/upgrade_guide.md [hunk 4]
```diff
@@ -22,6 +22,80 @@ $ kubectl edit deployment/etcd-operator
...
+    backupIntervalInSecond: 30
+    maxBackups: 5
+    storageType: "S3"
+  ```
+
+  to
+
+  ```yaml
+  backup:
+  backupIntervalInSecond: 30
```
Comment (hasbro17) [inline doc/user/upgrade/upgrade_guide.md]: Fix indent.

**Comment by fanminshi** [comment]
@hasbro17 still working on this pr. WIl ping once done.

**Comment by hasbro17** [comment]
Sure, guess I missed the WIP :)

**Comment by fanminshi** [comment]
All fixed PTAL

File: doc/user/upgrade/upgrade_guide.md [hunk 5]
```diff
@@ -22,6 +22,80 @@ $ kubectl edit deployment/etcd-operator
...
+### Breaking Change
+
+The v0.6.0 release removes [operator level s3 backup](https://github.com/coreos/etcd-operator/blob/v0.5.1/doc/user/backup_config.md#operator-level-configuration).
+If you are using operator level s3 backup in v0.5.x, you must migrate to [cluster level backup](../backup_config.md#s3-on-aws).
+
+Steps to upgrade:
+
+Assuming your cluster uses [operator level s3 backup](https://github.com/coreos/etcd-operator/blob/v0.5.1/doc/user/backup_config.md#operator-level-configuration).
+
+- Create a backup of you current data using the [backup config guide](../backup_config.md).
```
Comment (hasbro17) [inline doc/user/upgrade/upgrade_guide.md]: . . . using the [backup service guide](https://github.com/coreos/etcd-operator/blob/master/doc/user/backup_service.md#http-api-v1)

File: doc/user/upgrade/upgrade_guide.md [hunk 6]
```diff
@@ -22,6 +22,80 @@ $ kubectl edit deployment/etcd-operator
...
+- Update existing cluster:
+  `$ kubectl -n <namespace> apply -f <your-cluster-deployment>.yaml`
+
+- Update etcd-operator spec to use `etcd-operator:v0.6.0` and remove dependency on S3 level backup flags:
+
+  From
+
+  ```yaml
+  spec:
+      containers:
```
Comment (hasbro17) [inline doc/user/upgrade/upgrade_guide.md]: Probably best to keep the YAML indentation to 2 spaces for consistency.

File: doc/user/upgrade/upgrade_guide.md [hunk 7]
```diff
@@ -22,6 +22,80 @@ $ kubectl edit deployment/etcd-operator
...
+    storageType: "S3"
+    s3:
+      s3Bucket: <your-s3-bucket>
+      awsSecret: newAws
+  ```
+
+- Update existing cluster:
+  `$ kubectl -n <namespace> apply -f <your-cluster-deployment>.yaml`
+
+- Update etcd-operator spec to use `etcd-operator:v0.6.0` and remove dependency on S3 level backup flags:
```
Comment (hasbro17) [inline doc/user/upgrade/upgrade_guide.md]: remove dependency on S3 level backup flags => remove dependency on operator level S3 backup flags

File: doc/user/upgrade/upgrade_guide.md [hunk 8]
```diff
@@ -22,6 +22,81 @@ $ kubectl edit deployment/etcd-operator
...
 In the case of an incompatible upgrade, the process requires restoring a new cluster from backup. See the [incompatible upgrade guide](incompatible_upgrade.md) for more information.
 
+
+## v0.5.x -> v0.6.0
+
+### Breaking Change
+
+The v0.6.0 release removes [operator level s3 backup](https://github.com/coreos/etcd-operator/blob/v0.5.1/doc/user/backup_config.md#operator-level-configuration).
+
+If you are using operator level s3 backup, You must migrate your cluster to use [cluster level backup](../backup_config.md#s3-on-aws) for v0.6.0.
```
Comment (hongchaodeng) [inline doc/user/upgrade/upgrade_guide.md]: That's not true.
It's only required to migrate if the cluster spec uses S3 backup.

File: doc/user/upgrade/upgrade_guide.md [hunk 9]
```diff
@@ -22,6 +22,81 @@ $ kubectl edit deployment/etcd-operator
 ### Incompatible upgrade
 In the case of an incompatible upgrade, the process requires restoring a new cluster from backup. See the [incompatible upgrade guide](incompatible_upgrade.md) for more information.
 
+
+## v0.5.x -> v0.6.0
+
+### Breaking Change
+
+The v0.6.0 release removes [operator level s3 backup](https://github.com/coreos/etcd-operator/blob/v0.5.1/doc/user/backup_config.md#operator-level-configuration).
```
Comment (hongchaodeng) [inline doc/user/upgrade/upgrade_guide.md]: > operator level s3 backup 

operator S3 flag. Same for everywhere else.

File: doc/user/upgrade/upgrade_guide.md [hunk 10]
```diff
@@ -22,6 +22,81 @@ $ kubectl edit deployment/etcd-operator
...
+
+If you are using operator level s3 backup, You must migrate your cluster to use [cluster level backup](../backup_config.md#s3-on-aws) for v0.6.0.
+
+Steps to upgrade:
```
Comment (hongchaodeng) [inline doc/user/upgrade/upgrade_guide.md]: Please note that it's only needed if operator S3 flag is used.

File: doc/user/upgrade/upgrade_guide.md [hunk 11]
```diff
@@ -22,6 +22,81 @@ $ kubectl edit deployment/etcd-operator
...
+
+Assuming your cluster uses [operator level s3 backup](https://github.com/coreos/etcd-operator/blob/v0.5.1/doc/user/backup_config.md#operator-level-configuration).
```
Comment (hongchaodeng) [inline doc/user/upgrade/upgrade_guide.md]: Remove this line.

> Assuming your cluster uses 

Not a good manner to assume in upgrade docs.. Just tell users that they need to do something only if what condition is met.

File: doc/user/upgrade/upgrade_guide.md [hunk 12]
```diff
@@ -22,6 +22,81 @@ $ kubectl edit deployment/etcd-operator
...
+
+- Create a backup of you current data using the [backup service guide](../backup_service.md#http-api-v1).
+
+- Create a new AWS secret `newAws` for [cluster level backup](../backup_config.md#s3-on-aws):
```
Comment (hongchaodeng) [inline doc/user/upgrade/upgrade_guide.md]:
```
The Secret "newAws" is invalid: metadata.name: Invalid value: "newAws": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

Please try the docs and make sure it works.

Comment (hongchaodeng) [reply doc/user/upgrade/upgrade_guide.md]: Also it is not a good way to give a specific name here.

File: doc/user/upgrade/upgrade_guide.md [hunk 13]
```diff
@@ -22,6 +22,81 @@ $ kubectl edit deployment/etcd-operator
...
+    storageType: "S3"
+    s3:
+      s3Bucket: <your-s3-bucket>
+      awsSecret: newAws
+  ```
+
+- Update existing cluster:
+  `$ kubectl -n <namespace> apply -f <your-cluster-deployment>.yaml`
+
+- Update etcd-operator spec to use `etcd-operator:v0.6.0` and remove dependency on operator level S3 backup flags:
```
Comment (hongchaodeng) [inline doc/user/upgrade/upgrade_guide.md]: Update ... deployment spec

File: doc/user/upgrade/upgrade_guide.md [hunk 14]
```diff
@@ -22,6 +22,81 @@ $ kubectl edit deployment/etcd-operator
...
+  ```yaml
+  spec:
+    containers:
+    - name: etcd-operator
+      image: quay.io/coreos/etcd-operator:v0.6.0
+      command: 
+        - /usr/local/bin/etcd-operator
+  ```
+
+- Update existing etcd-operator deployment:
```
Comment (hongchaodeng) [inline doc/user/upgrade/upgrade_guide.md]: This is duplicate to above point. Remove this?

File: doc/user/upgrade/upgrade_guide.md (hunk 8 above)
Comment (fanminshi) [reply doc/user/upgrade/upgrade_guide.md]: I should be more clear.

I want to say that if you are using  operator level s3 backup and want to enable backup in v0.6.0, then you need to migrate your cluster to use [cluster level backup].

File: doc/user/upgrade/upgrade_guide.md (hunk 14 above)
Comment (fanminshi) [reply doc/user/upgrade/upgrade_guide.md]: the above ask you to update the config. this line ask you to apply the config change

**Comment by fanminshi** [comment]
questions addressed

**Comment by hongchaodeng** [comment]
lgtm
Great work! @fanminshi

**Comment by hongchaodeng** [comment]
fix https://github.com/coreos/etcd-operator/issues/1377

**Comment by xiang90** [comment]
if we are going to merge in https://github.com/coreos/etcd-operator/pull/1400, then it is another breaking change.

**Comment by hongchaodeng** [comment]
Let's merge this first.
#1400 is still undecided.

**Pull request merged**